oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/MappedJwtClaimSetConverter.java

```diff
@@ -0,0 +1,288 @@
+
+		IssuedAtConverter issuedAtConverter = new IssuedAtConverter();
+
+		Map<String, Converter<Map<String, Object>, ?>> withDefaults = new HashMap<>();
+		withDefaults.put(JwtClaimNames.AUD, new AudienceConverter());
+		withDefaults.put(JwtClaimNames.EXP, new TemporalConverter(JwtClaimNames.EXP));
+		withDefaults.put(JwtClaimNames.IAT, issuedAtConverter);
+		withDefaults.put(JwtClaimNames.ISS, new IssuerConverter());
+		withDefaults.put(JwtClaimNames.JTI, new StringConverter(JwtClaimNames.JTI));
+		withDefaults.put(JwtClaimNames.NBF, new TemporalConverter(JwtClaimNames.NBF));
+		withDefaults.put(JwtClaimNames.SUB, new StringConverter(JwtClaimNames.SUB));
+		withDefaults.putAll(attributeConverters);
+
+		MappedJwtClaimSetConverter converter = new MappedJwtClaimSetConverter(withDefaults);
+		issuedAtConverter.setAttributeConverters(converter.attributeConverters);
+
+		return converter;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public Map<String, Object> convert(Map<String, Object> claims) {
+		Assert.notNull(claims, "claims cannot be null");
+
+		Map<String, Object> mappedClaims = new HashMap<>(claims);
+
+		for (Map.Entry<String, Converter<Map<String, Object>, ?>> entry : this.attributeConverters.entrySet()) {
+			Converter<Map<String, Object>, ?> converter = this.attributeConverters.get(entry.getKey());
+			if (converter != null) {
+				put(mappedClaims, entry.getKey(), converter.convert(mappedClaims));
+			}
+		}
+
+		return mappedClaims;
+	}
+
+	private static class AudienceConverter implements Converter<Map<String, Object>, Collection<String>> {
+
+		@Override
+		public Collection<String> convert(Map<String, Object> source) {
+			Object audience = source.get(JwtClaimNames.AUD);
+
+			if (audience == null) {
+				return null;
+			}
+
+			if (audience instanceof Collection) {
+				return ((Collection<?>) audience).stream()
+						.filter(Objects::nonNull)
+						.map(Objects::toString)
+						.collect(Collectors.toList());
+			}
+
+			return Arrays.asList(audience.toString());
+		}
+	}
+
+	private static class IssuerConverter implements Converter<Map<String, Object>, URL> {
+
+		@Override
+		public URL convert(Map<String, Object> source) {
+			Object issuer = source.get(JwtClaimNames.ISS);
+
+			if (issuer == null) {
+				return null;
+			}
+
+			if (issuer instanceof URL) {
+				return (URL) issuer;
+			}
+
+			if (issuer instanceof URI) {
+				return toUrl((URI) issuer);
+			}
+
+			return toUrl(issuer.toString());
+		}
+
+		private URL toUrl(String issuer) {
+			try {
+				return new URL(issuer);
+			} catch (MalformedURLException e) {
+				throw new IllegalStateException("Could not coerce iss value to URL");
+			}
+		}
+
+		private URL toUrl(URI issuer) {
+			try {
+				return issuer.toURL();
+			} catch (MalformedURLException e) {
+				throw new IllegalStateException("Could not coerce iss value to URL");
+			}
+		}
+	}
+
+	private static class IssuedAtConverter extends TemporalConverter {
```

Comment:
Favor composition to inheritance since Java only lets you inherit from a single class. Instead the temporal converter could accept an actual value (instead of the entire Map).